hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,179 @@
+        return;
+      }
+      running = true;
+      TableDescriptors htds = master.getTableDescriptors();
+      Map<String, TableDescriptor> map = htds.getAll();
+      for (TableDescriptor htd : map.values()) {
+        if (!master.getTableStateManager().isTableState(htd.getTableName(),
+          TableState.State.ENABLED)) {
+          continue;
+        }
+        for (ColumnFamilyDescriptor hcd : htd.getColumnFamilies()) {
+          if (hcd.isMobEnabled()) {
+            if (!reported) {
+              master.reportMobCompactionStart(htd.getTableName());
+              reported = true;
+            }
+            LOG.info(" Major compacting "+ htd.getTableName() + " cf=" + hcd.getNameAsString());
+            if (regionBatchSize == MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE) {
+              admin.majorCompact(htd.getTableName(), hcd.getName());
+            } else {
+              performMajorCompactionInBatches(admin, htd, hcd);
+            }
+          }
+        }
+        if (reported) {
+          master.reportMobCompactionEnd(htd.getTableName());
+          reported = false;
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to compact", e);
+    } finally {
+      running = false;
+    }
+  }
+
+  private void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i=0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for(RegionInfo ri: toCompact) {
+      startCompaction(admin, htd.getTableName(), ri);
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    while(!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri: toCompact) {
+        if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+          LOG.info("Finished major compaction: table={} region={}", htd.getTableName(),
+            ri.getRegionNameAsString());
+          compacted.add(ri);
+        }
+      }
+      // Update batch: remove compacted regions and add new ones
+      for (RegionInfo ri: compacted) {
+        toCompact.remove(ri);
+        if (regions.size() > 0) {
+          RegionInfo region = regions.remove(0);
+          startCompaction(admin, htd.getTableName(),region);
+          toCompact.add(region);
+        }
+      }
+      compacted.clear();
+      try {
+        Thread.sleep(1000);
+      } catch(InterruptedException e) {
+        // swallow
```

Comment:
why are we swallowing instead of being interrupted? if we receive an interruption shouldn't we empty toCompact and set the interrupted state?

Reply:
The new code propagates InterruptedException up to the chore() method, which restores interrupted states, report master compaction ends and exits the chore. 

```
catch (InterruptedException ee) {
            Thread.currentThread().interrupt();
            master.reportMobCompactionEnd(htd.getTableName());
            LOG.warn("Failed to compact table="+ htd.getTableName() +" cf="+ hcd.getNameAsString(),
              ee);
            // Quit the chore
            return;
          }
```
Let me know what you think.

Reply:
yeah that sounds fine.